src/ucp/dt/datatype_iter.inl

```diff
@@ -195,18 +161,21 @@ ucp_datatype_iter_init(ucp_context_h context, void *buffer, size_t count,
                        uint8_t *sg_count, const ucp_request_param_t *param)
 {
     size_t iov_length;
+    int detect_memory;
 
     dt_iter->offset   = 0;
     dt_iter->dt_class = ucp_datatype_class(datatype);
+    detect_memory = !ucp_datatype_iter_is_memh_supplied(param->op_attr_mask,
+                                                        dt_iter->dt_class);
 
     if (ucs_likely(dt_iter->dt_class == UCP_DATATYPE_CONTIG)) {
-        ucp_datatype_contig_iter_init(context, buffer, contig_length, datatype,
-                                      dt_iter);
+        ucp_datatype_contig_iter_init(context, buffer, contig_length, dt_iter,
```

Comment:
ucp_datatype_contig_iter_init/ucp_datatype_iov_iter_init should get "param" as parameter and init memtype accordingly. this way can avoid checking dt_class in ucp_datatype_iter_is_memh_supplied and ucp_datatype_iter_set_memh

Reply:
notice that `ucp_datatype_iter_init_from_dt_state` is also calling `ucp_datatype_contig_iter_init`/`ucp_datatype_iov_iter_init`,
but this function does not know param.

Reply:
it can pass some dummy param struct from global var where field_mask is 0

Reply:
regarding `ucp_datatype_iter_set_memh`, it's not inside the scope of `ucp_datatype_contig_iter_init/ucp_datatype_iov_iter_init`, so we must check dt_class

Reply:
need to move the logic of ucp_datatype_iter_set_memh inside ucp_datatype_contig_iter_init/ucp_datatype_iov_iter_init

Reply:
you mean completely remove this function? (and move it's content to the other functions?)